src/Components/Web.JS/src/Rendering/BrowserRenderer.ts

```diff
@@ -220,7 +220,17 @@ export class BrowserRenderer {
   private insertElement(batch: RenderBatch, componentId: number, parent: LogicalElement, childIndex: number, frames: ArrayValues<RenderTreeFrame>, frame: RenderTreeFrame, frameIndex: number) {
     const frameReader = batch.frameReader;
     const tagName = frameReader.elementName(frame)!;
-    const newDomElementRaw = tagName === 'svg' || isSvgElement(parent) ?
+    
+    // SVG elements support `foreignObject` children that can hold arbitrary HTML.
+    // For these scenarios, the parent SVG and `foreignObject` elements should
+    // be rendered under the SVG namespace, while the HTML content should be rendered
+    // under the XHTML namespace. If the correct namespaces are not provided, most
+    // browsers will fail to render the foreign object content. Here, we ensure that if
+    // we encounter a `foreignObject` in the SVG, then all its children will be placed
+    // under the XHTML namespace.
+    const isSvg = (tagName === 'svg' || isSvgElement(parent)) && !isForeignObject(parent);
```

Comment:
Note that this isn't the only place where we check `isSvgElement(parent)`. We also do that in `insertMarkup` to vary how we parse markup strings. My guess is that we'd also need to update `insertMarkup` in the same way as this in case we're inserting a markup block as the top-level thing inside a foreign element. Does that sound correct to you?

If this is all true, it suggests to me that instead of having both `isSvgElement` and `isForeignObject` as separate things we have to remember to use together, we should instead replace them both with a single `shouldTreatImmediateChildrenAsSvg` function. This would also have an efficiency benefit, since that function would only need to evaluate `getClosestDomElement(element)` once instead of two methods both evaluating that.

Reply:
Great catch!

Yep -- it makes sense to refactor things a bit with an efficiency focus if these check happens more than once.

Will update PR accordingly.